zebra-utils/src/bin/zebra-checkpoints/main.rs

```diff
@@ -1,32 +1,108 @@
+//! If no network is specified program will use the mainnet.
+//!
+//! zebra-consensus accepts an ordered list of checkpoints, starting with the
+//! genesis block. Checkpoint heights can be chosen arbitrarily.
+
 #![allow(clippy::try_err)]
-use color_eyre::eyre::{eyre, Result};
+use color_eyre::eyre::Result;
+use serde_json::Value;
+use std::process::Stdio;
 use structopt::StructOpt;
 
 mod args;
 
+/// We limit the maximum number of blocks in each checkpoint. Each block uses a
+/// constant amount of memory for the supporting data structures and futures.
+const MAX_CHECKPOINT_HEIGHT_GAP: usize = zebrad::commands::LOOKAHEAD_LIMIT / 2;
+
+/// We limit the memory usage for each checkpoint, based on the cumulative size of
+/// the serialized blocks in the chain. Deserialized blocks are larger, because
+/// they contain pointers and non-compact integers. But they should be within a
+/// constant factor of the serialized size.
+const MAX_CHECKPOINT_BYTE_COUNT: i64 = 256 * 1024 * 1024;
+
+/// Checkpoints must be on the main chain, so we skip blocks that are within the
+/// zcashd reorg limit.
+const BLOCK_REORG_LIMIT: i64 = 100;
+
+// Add network argument if needed
+fn network(mut cmd: std::process::Command, args: &args::Args) -> std::process::Command {
+    if args.testnet {
+        cmd.arg("-testnet");
+    } else if args.regtest {
+        cmd.arg("-regtest");
+    }
+    cmd
+}
+
 fn main() -> Result<()> {
     // todo add tracing setup
 
     color_eyre::install()?;
 
+    // create process
     let args = args::Args::from_args();
+    let mut cmd = std::process::Command::new(&args.cli);
+    cmd = network(cmd, &args);
 
-    let mut cmd = std::process::Command::new("zcash-cli");
+    // set up counters
+    let mut cumulative_bytes: i64 = 0;
+    let mut height_gap: usize = 0;
 
-    if args.testnet {
-        cmd.arg("-testnet");
-    }
+    // get the current block count
+    cmd.arg("getblockcount");
+    cmd.stdout(Stdio::piped()).spawn().unwrap();
+    let output = cmd.output().unwrap();
+    let mut block_count: i64 = String::from_utf8_lossy(&output.stdout)
+        .trim()
+        .parse()
+        .unwrap();
+    block_count -= BLOCK_REORG_LIMIT;
+
+    // loop through all blocks
+    for x in 0..block_count {
+        // unfortunatly we need to create a process for each block
+        let mut cmd = std::process::Command::new(&args.cli);
+        cmd = network(cmd, &args);
+
+        // get block data
+        cmd.args(&["getblock", &x.to_string()]);
+        cmd.stdout(Stdio::piped()).spawn().unwrap();
+        let output = cmd.output().unwrap();
+        let block_raw = String::from_utf8_lossy(&output.stdout);
 
-    cmd.args(args.zcli_args.into_iter());
+        // convert raw block to json
+        let v: Value = serde_json::from_str(block_raw.trim())?;
 
-    let mut child = cmd.spawn()?;
+        // get the values we are interested in
+        let hash = &v["hash"].as_str().unwrap();
```

Comment:
```suggestion
        let hash: BlockHeaderHash = &v["hash"].as_str().unwrap().map(zebra_chain::utils::byte_reverse_hex).parse();
```

Reply:
I think this will be good but provided code don't work as it is and i was unable to build a good replacement.

Reply:
```suggestion
        let hash: BlockHeaderHash = v["hash"]
            .as_str()
            .map(zebra_chain::utils::byte_reverse_hex)
            .unwrap()
            .parse()
            .unwrap();
```

should work

Reply:
Eeh, I can't repro the weird errors I ran into, lame. :(

Reply:
None of the 2 works actually because `byte_reverse_hex` is expecting a string. I think the conversion to `BlockHeaderHash` might not be needed here unless we want to do checks on the data of it.

I guess we could change the `byte_reverse_hex` signature.

Reply:
Are you sure neither of the suggestions I posted work? I pulled down your commit and they were working for me. `as_str()` has a return type of `Option<&str>` and `bytes_reverse_hex` has a signature of `fn(&str) -> String` so this should match up and work directly in `map`.

Reply:
Sorry, i got confused with another line failing after the change that haves nothing to do with this :(

Reply:
Ok, apart from my confusion i do think doing the `bytes_reverse_hex` inline is a good idea but i don't see the need of converting it to `BlockHeaderHash` as this will imply converting it to string again at the end:  https://github.com/oxarbitrage/zebra/blob/issue684/zebra-utils/src/bin/zebra-checkpoints/main.rs#L109

I did implemented the map version but leaving the variable as a string in https://github.com/ZcashFoundation/zebra/pull/740/commits/2f1109a13628a27aceba9a1a6c7bcf9aeb4951f5

Reply:
> Ok, apart from my confusion i do think doing the `bytes_reverse_hex` inline is a good idea but i don't see the need of converting it to `BlockHeaderHash` as this will imply converting it to string again at the end

One advantage is that it validates that the input is valid hex

```rust
impl std::str::FromStr for BlockHeaderHash {
    type Err = SerializationError;
    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut bytes = [0; 32];
        if hex::decode_to_slice(s, &mut bytes[..]).is_err() {
            Err(SerializationError::Parse("hex decoding error"))
        } else {
            Ok(BlockHeaderHash(bytes))
        }
    }
}
```

I'm not sure this is actually valuable though so I'm okay with not including it.

Reply:
We're getting the hash from an outside program, so we need to check that it is the right length, and that it is actually a hex string.

Converting to BlockHeaderHash is an easy and reliable way to do that.

Reply:
Ok, fair enough. i added this at https://github.com/ZcashFoundation/zebra/pull/740/commits/695d6cda83f5f1192956ffd7140ced1632249b06